src/compiler/checker.ts

```diff
@@ -43537,7 +43536,7 @@ namespace ts {
                 flags |= NodeBuilderFlags.AllowUniqueESSymbolType;
             }
             if (addUndefined) {
-                type = getOptionalType(type);
+                type = getOptionalType(type, /*isProperty*/ undefined, /*ignoreStrictNullChecks*/ true);

```

Comment:
I think a better way to fix this would be to temporarily set `strictNullChecks = true` and then run `getOptionalType`, then reset it afterward. @weswigham does that make sense to you? Or can you suggest a different way entirely?

Reply:
Any and every initialized parameter property needs a `?` if it's a trailing parameter. This is something that the transformer needs to check, beyond the `resolver.isOptionalParameter` check it already does. Once that's handled, rather than doing anything strict-null-check-dependent, I'd keep it syntactic, and likely in the declaration emitter. If the parameter _doesn't_ get a `?` (because it has a required parameter after it), then we add `| undefined` to the type, regardless of mode, provided the existing serialized type doesn't contain it. (We don't need/want to do this extra explicit `undefined` addition for stuff with the `?` so we can better strict optional property check them, I think.)

Reply:
@weswigham Very insightful! Should just think outside the box. Now the implementation is simplified and several bugs are fixed.